Add Rustiq-based synthesis for `PauliEvolutionGate`

alexanderivrii opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

Calls the `Rustiq's` synthesis algorithm for pauli networks natively from within Qiskit. 

Together with #13295 addresses #12789.
This should be rebased on top of #13295, in order to support the plugin interface structure for `PauliEvolutionGate`s.

Rustiq is available at https://github.com/smartiel/rustiq-core


### How to experiment

One can experiment with the new functionality by creating a `PauliEvolutionGate` and directly calling the new `PauliEvolutionSynthesisRustiq` on this gate:

```
from qiskit.transpiler.passes.synthesis.hls_plugins import PauliEvolutionSynthesisRustiq

terms = ["IX", "ZZ"]
evo_gate = PauliEvolutionGate(SparsePauliOp(terms), time=0.12)
qct = PauliEvolutionSynthesisRustiq().run(evo_gate)
```

Alternatively, one can create a circuit with `PauliEvolutionGate`s and specify "rustiq" via the synthesis config interface:

```
terms = ["IX", "ZZ"]
evo_gate = PauliEvolutionGate(SparsePauliOp(terms), time=0.12)
qc = QuantumCircuit(3)
qc.append(evo_gate, [1, 2])
config = HLSConfig(PauliEvolution=["rustiq"])
qct = transpile(qc, hls_config=config)
```

### Details:

The Qiskit Rust function `pauli_network_synthesis` calls the `greedy_pauli_network` function from the `rustiq-core` (in particular, most of the code is fully in Rust). However, the latter function only returns a sequence of Clifford gates, without the single-qubit rotations. So `pauli_network_synthesis` also adapts the code from Simon's private rustiq-plugin repository that simulates the circuit and inserts rotations wherever required.

### Preliminary Experiments:

To be updated (or maybe it would be even best to move this discussion to the issue #12789). After running some preliminary experiments on some specific examples (e.g. see #13285) and some hamlib examples (taken from https://github.com/Qiskit/benchpress/blob/main/benchpress/hamiltonian/hamlib/), `rustiq` can be amazingly powerful in many cases, greatly reducing the two-qubits gate-count and depth of the resulting circuit. On the other hand, there are also cases where Rustiq performs worse, i.e. creating a circuit with more gates.

### Some Concerns:

- On some of the hamlib examples Rustiq is extremely fast, yet there are also examples where it takes more than 1 hour to complete. 
- In some cases the resulting circuit is not fully deterministic. I am currently trying to understand why this happens. 
- So far in all the cases the circuit produced by Rustiq is equivalent to the circuit produced by the default synthesis algorithm in Qiskit, _up to the global phase_, however the global phase difference is not always 0. This is after taking care of global phase updates related to all-I pauli rotation gates. I am also trying to understand why this happens.

It is possible that some additional changes to Rustiq itself might be required. 

### ToDo:

- [x] Docstring improvements, release notes, new plugin documentation, documenting rustiq-specific options, etc. 
- [x] Tests: make sure that the output circuits are correct and are equivalent to the circuits produced by the default algorithm, make sure that the plugins run as intended

Cryoris committed:
py version for expand

alexanderivrii committed:
starting to write some code

alexanderivrii committed:
merge main

alexanderivrii committed:
implementing

alexanderivrii committed:
cleanup

alexanderivrii committed:
cleanup

Cryoris committed:
Merge branch 'main' into paulievo

Cryoris committed:
expand fully & simplify lie trotter

Cryoris committed:
use examples that actually do not commute

Cryoris committed:
add plugin structure

alexanderivrii committed:
fixing global phase for all-I rotations

alexanderivrii committed:
Merge branch 'paulievo' into integrate_rustiq_plugin

alexanderivrii committed:
fixes

alexanderivrii committed:
fixing plugin names

alexanderivrii committed:
Merge branch 'main' into integrate_rustiq_plugin

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,336 @@
+///   * should we allow all-I rotations?
+
+/// A Qiskit gate
+pub type QiskitGate = (StandardGate, SmallVec<[Param; 3]>, SmallVec<[Qubit; 2]>);
```
Does this need to be public? 🙂

alexanderivrii replied:
fixed!

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,336 @@
+/// For example: for the input `sparse_pauli = "XY", qubits = [1, 3], num_qubits = 6`,
+/// the function returns `"IXIYII"`.
+fn expand_pauli(sparse_pauli: String, qubits: Vec<u32>, num_qubits: usize) -> String {
+    let mut v: Vec<char> = vec!["I".parse().unwrap(); num_qubits];
```
Does this also work? That should directly use the `char` type
```suggestion
    let mut v: Vec<char> = vec!['I'; num_qubits];
```

alexanderivrii replied:
Great, indeed!

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,336 @@
+            return (standard_gate, smallvec![param], smallvec![Qubit(q as u32)]);
+        }
+    }
+    unreachable!()
```
It might be nice to add an explanation as to why this is unreachable, e.g.
```suggestion
    unreachable!("The paulis are guaranteed to have at least one non-I term at this point.")
```

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,336 @@
+
+impl CommutativityDag {
+    /// Construct a DAG based on the commutativity relations between paulis.
+    pub fn from_paulis(paulis: &PauliSet) -> Self {
```
Same question here, do these methods need to be public? Since the struct itself is not I would assume that we can just have these private? 🤔

alexanderivrii replied:
yes, fixed

alexanderivrii committed:
minor

alexanderivrii committed:
Merge branch 'integrate_rustiq_plugin' of github.com:alexanderivrii/qiskit-terra into integrate_rustiq_plugin

Cryoris commented on crates/accelerate/src/synthesis/permutation/mod.rs:
```diff
@@ -37,6 +37,7 @@ pub fn _validate_permutation(py: Python, pattern: PyArrayLike1<i64>) -> PyResult
 #[pyfunction]
 #[pyo3(signature = (pattern))]
 pub fn _inverse_pattern(py: Python, pattern: PyArrayLike1<i64>) -> PyResult<PyObject> {
+    println!("I AM HERE YES");
```
🖨️

alexanderivrii committed:
removing a random print statement

Cryoris commented on qiskit/synthesis/evolution/__init__.py:
```diff
@@ -12,6 +12,9 @@
 """Synthesis for operator evolution gates."""
 
+# pylint: disable=unused-import
+from qiskit._accelerate.synthesis.evolution import pauli_network_synthesis
```
The classes in here perform the full synthesis of a `PauliEvolutionGate` to a `QuantumCircuit`,  which this function does not. Would it make more sense to expose a similar wrapper here (basically what the plugin is doing) or just to remove this function as it is more a utility for the plugin?

alexanderivrii replied:
I agree. To be consistent with the other synthesis methods, I have wrapped this into a function `synth_pauli_network_rustiq`.

Cryoris reviewed (commented):
I just had a brief look, but overall this looks good to me -- I'll give a detailed review later 🙂

alexanderivrii committed:
additional improvements

alexanderivrii committed:
improving rustiq plugin

alexanderivrii committed:
Merge branch 'main' into integrate_rustiq_plugin

alexanderivrii committed:
Merge branch 'main' into integrate_rustiq_plugin

alexanderivrii committed:
Merge branch 'main' into integrate_rustiq_plugin

alexanderivrii committed:
merge with #13239

alexanderivrii committed:
Adding pauli evolution plugins to docstrings

alexanderivrii committed:
adding documentation on rustiq plugin

alexanderivrii committed:
fixes after refactoring

alexanderivrii committed:
typo

alexanderivrii committed:
more merges with #13295; adding more Rustiq tests

alexanderivrii committed:
Merge branch 'main' into integrate_rustiq_plugin

ShellyGarion commented on crates/accelerate/src/synthesis/clifford/utils.rs:
```diff
@@ -227,6 +227,24 @@ impl Clifford {
+                    Ok(())
+                }
+                StandardGate::SXGate => {
+                    // todo: replace by a single update
```
this can be done already in this PR, see:
https://github.com/Qiskit/qiskit/blob/db0433920c86e55f4ec06b3837d10d94b040e117/qiskit/quantum_info/operators/symplectic/clifford_circuits.py#L325

alexanderivrii replied:
Thanks, I planned to do it, but forgot. Added in [41ae2f3](https://github.com/Qiskit/qiskit/pull/13301/commits/41ae2f33146762e3ec0b0e456d3f63dc40e0d0c9).

ShellyGarion commented on crates/accelerate/src/synthesis/clifford/utils.rs:
```diff
@@ -227,6 +227,24 @@ impl Clifford {
+                    Ok(())
+                }
+                StandardGate::SXdgGate => {
+                    // todo: replace by a single update
```
this can be done already as part of this PR, see:
https://github.com/Qiskit/qiskit/blob/db0433920c86e55f4ec06b3837d10d94b040e117/qiskit/quantum_info/operators/symplectic/clifford_circuits.py#L343

alexanderivrii replied:
Same here. Added in [41ae2f3](https://github.com/Qiskit/qiskit/pull/13301/commits/41ae2f33146762e3ec0b0e456d3f63dc40e0d0c9).

ShellyGarion commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,431 @@
+// some point to a single-qubit rotation, in the given order, but up to commutativity.
+// Following the code in Simon's private rustiq-plugin repository, we simulate the
+// clifford gates to find where pauli rotations need to be inserted. If the future
+// the synthesis algorithm will explicitly return where the rotations need to be
```
If the future the synthesis ? --> remove one "the"

alexanderivrii replied:
I changed the wording of this entire note to explain more clearly the challenges of integrating rustiq code.

ShellyGarion commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,431 @@
+use qiskit_circuit::Qubit;
+
+use rustiq_core::structures::{
+    CliffordCircuit, CliffordGate, IsometryTableau, Metric, PauliLike, PauliSet,
```
Is there a check needed that the user installed rustiq before using this code and raise a warning otherwise?

alexanderivrii replied:
No, as I wrote somewhere else, the users do not need to install anything. Qiskit will simply not compile if the above package does not exist. Also note that in `Cargo.toml` we explicitly pin rustiq's version to "0.0.8", so we are also safe against any updates to rustiq-core.

ShellyGarion commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,431 @@
+///     generally not be equivalent to the given pauli network.
+/// * upto_clifford: if `true`, the final Clifford operator is not synthesized
+///     and the returned circuit will generally not be equivalent to the given
+///     pauli network. In addition, the argument `upto_phase` would be ignored.
```
If it's the final Clifford, then one can have both "upto_clifford" and "upto_phase" since the global phase should not affect the final measurement?

alexanderivrii replied:
I don't know if it makes sense to ignore the final Clifford but to keep the global phase. And in any case currently we don't have a way to do that (without the extended Clifford simulator). So the assumption that I tried to document here is that if we ignore the final Clifford, we also automatically ignore the global phase. Anyway, for now this is really an experimental option available to users (I am pretty sure that @ajavadia would like to have it). I also expect that in future, when synthesizing up to the final Clifford, the calling application would like to obtain this Clifford as a part of the result (for instance, to modify the measurement operators that need to be performed), and we currently don't have API for any of this.

ShellyGarion commented on qiskit/synthesis/evolution/pauli_network.py:
```diff
@@ -0,0 +1,80 @@
+    Calls Rustiq's pauli network synthesis algorithm.
+
+    The algorithm is described in [1]. The source code (in Rust) is available at
+    https://github.com/smartiel/rustiq-core.
```
should the user install and compile rustiq-core for this code ?

alexanderivrii replied:
No, the rustiq-core (version 0.0.8) is available [crates.io](https://crates.io/crates/rustiq-core), and it's "pulled in" during Qiskit's Rust compilation (as all the other Rust crates Qiskit internally uses). There is nothing specific a user needs to install.

ShellyGarion reviewed (commented):
LGTM. I only have a few comments and questions.

ShellyGarion commented on qiskit/transpiler/passes/synthesis/hls_plugins.py:
```diff
@@ -1047,8 +1079,88 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
+        commutativity.
+    * upto_clifford (bool): if `True`, the final Clifford operator is not synthesized.
+    * upto_phase (bool): if `True`, the global phase of the returned circuit may
+        differ from the global phase of the given pauli network.
```
so according to the docs, one can have both "upto_clifford=True" and "upto_phase=True" ?

alexanderivrii replied:
Yes, so if both are `False`, we want the exact synthesis -- this is the default behavior. If `upto_clifford=False` but `upto_phase` is `True`, then we can synthesize ignoring the global phase, in which case by default we use Qiskit's Clifford resynthesis, since it's usually faster and gives a very comparable quality of results. If `upto_clifford` it `True` then (as noted somewhere else) we don't care about the global phase either.

alexanderivrii committed:
more efficient append_sx and append_sxdg gates for cliffords

alexanderivrii committed:
review comments

alexanderivrii committed:
Merge branch 'integrate_rustiq_plugin' of github.com:alexanderivrii/qiskit-terra into integrate_rustiq_plugin

ShellyGarion reviewed (dismissed):
LGTM - thanks!

Cryoris commented on crates/accelerate/src/synthesis/clifford/utils.rs:
```diff
@@ -149,6 +149,30 @@ impl Clifford {
     }
 
+    /// Modifies the tableau in-place by appending SX-gate
+    pub fn append_sx(&mut self, qubit: usize) {
```
These look correct, so I'm fine adding as is, but do we have a test that could specifically trigger these? Like constructing a Clifford object from a circuit with SX gates? 🙂

alexanderivrii replied:
In general, the Rust-based Clifford code is implicitly tested when checking the correctness of synthesis results in `test_high_level_synthesis.py`, etc.. However, this is not a full-proof method since this only considers gates produced by the synthesis algorithms. This is why we did not need handling `SX` and `SXsg` gates before this PR, none of our current Clifford synthesis algorithms produced such gates. 

The best way I can think of how to apply your suggestion is to add Rust-based tests that we would then execute with `cargo test`. However, so far we have been mostly avoiding this path. Note that our internal Rust-based Clifford code is not exposed via a Python interface, this is why I don't see how to write directed tests from the Python space. Would love to hear additional suggestions.

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
Could we move the logic into a separate file, that is not `mod.rs`? I think in most code, the `mod.rs` only has the module imports or, at most, the `pyfunction` definition and not much logic. Otherwise we might end up with very different functionalities in the asme file 🙂

alexanderivrii replied:
Done (the commit will soon follow).

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+#![allow(clippy::too_many_arguments)]
```
I would assume theres only very few functions that need this exception, could you move it to these instead of a module-level except? 🙂

alexanderivrii replied:
I though that the module-level except was the preferred way since this is how we have it in `dense_layout.rs` and many other files. But, I agree, for a single function the function-level except makes more sense. Done (the commit will soon follow).

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+///
+/// For example: for the input `sparse_pauli = "XY", qubits = [1, 3], num_qubits = 6`,
+/// the function returns `"IXIYII"`.
+fn expand_pauli(sparse_pauli: String, qubits: Vec<u32>, num_qubits: usize) -> String {
```
Could this take a qubit slice (i.e. `qubits: &[u32]`) to avoid copying the vector itself?

alexanderivrii replied:
Ah, in my code I was passing the ownership to `qubits` when calling `expand_pauli`, which compiled since `qubits` were not used after that. But agree that passing this as a slice is a far better API. Done (the commit will soon follow).

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+fn expand_pauli(sparse_pauli: String, qubits: Vec<u32>, num_qubits: usize) -> String {
+    let mut v: Vec<char> = vec!['I'; num_qubits];
+    for (i, q) in qubits.iter().enumerate() {
+        v[*q as usize] = sparse_pauli.chars().nth(i).unwrap();
```
Just a suggestion, as I don't know if `chars().nth()` is for free
```suggestion
    for (q, p) in qubits.iter().zip(sparse_pauli.chars()) {
        v[*q as usize] = p;
```
(this might require some dereferencing or cloning -- my rust skills decrease by 98% without `rust-analyzer`)

alexanderivrii replied:
This works! I also don't know how expensive `chars().nth()` really is (after all, both approaches should just boil down to a single memory access), but your suggestion is visually better. Changed (the commit will soon follow).

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+/// * py: a GIL handle, needed to negate rotation parameters in Python space.
+/// * paulis: Rustiq's data structure storing pauli rotations.
+/// * i: index of the single-qubit Pauli rotation.
+/// * angle: - Qiskit's rotation angle.
```
```suggestion
/// * angle: Qiskit's rotation angle.
```

alexanderivrii replied:
Changed (the commit will soon follow).

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+                'X' => StandardGate::RXGate,
+                'Y' => StandardGate::RYGate,
+                'Z' => StandardGate::RZGate,
+                _ => panic!(),
```
Personally, I would use `unreachable` here as we don't expect this arm can be reached, whereas `panic` is maybe less precise? Plus it might be nice adding a small message to simplify tracking down where this came from 🙂 But I'm also happy to leave as is, this is just a suggestion.
```suggestion
                _ => unreachable!("Only X, Y and Z are possible characters at this point."),
```

alexanderivrii replied:
Changed (the commit will soon follow).

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+/// * angle: - Qiskit's rotation angle.
+fn qiskit_rotation_gate(py: Python, paulis: &PauliSet, i: usize, angle: &Param) -> QiskitGate {
+    let (phase, pauli_str) = paulis.get(i);
+    for (q, c) in pauli_str.chars().enumerate() {
```
Is `pauli_str` guaranteed to be upper case? If not, it might be worth calling `.upper()` on it first

alexanderivrii replied:
Yes (and I double-checked rustiq-core's `PauliSet::get` function to confirm).

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+    fn from_paulis(paulis: &PauliSet) -> Self {
+        let mut dag = StableDiGraph::<usize, ()>::new();
+
+        let node_indices: Vec<NodeIndex> = (0..paulis.len()).map(|i| dag.add_node(i)).collect();
```
I'm surprised clippy doesn't complain about this, but is this equivalent?
```suggestion
        let node_indices: Vec<NodeIndex> = (0..paulis.len()).map(dag.add_node).collect();
```

alexanderivrii replied:
Hmm, the suggested made perfect sense to me, and I am sure I've used something similar in the past, but in fact this does not compile with the error that `dag.add_node` is a method and not a field. I guess if the function `add_node` was not a part of the DAG class, this would have worked.

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+/// * angles: Qiskit's rotation angles corresponding to the pauli rotations.
+fn inject_rotations_unordered(
+    py: Python,
+    gates: &Vec<CliffordGate>,
```
From what the internet tells me "using &Vec over &[..] implies an extra pointer re-direction" and I remember clippy telling me I should use the slice over a vector -- but I'm fine as is if clippy is fine 🙂

alexanderivrii replied:
I am not sure whether `&[..]` requires less dereferencing than `&Vec<..>`, however it definitely works for a larger number of structures. Now that you mentioned it, I would have expected clippy to complain, I have seen similar complains in the past. Done (the commit will soon follow).

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+///     pauli network synthesis algorithm.
+/// * paulis: Rustiq's data structure storing the pauli rotations.
+/// * angles: Qiskit's rotation angles corresponding to the pauli rotations.
+fn inject_rotations_ordered(
```
Does this only differ from the `unordered` version in the few lines that create/modify the commutativity dag? If yes, would it make sense to merge them and add a keyword `ordered: bool` to avoid code duplication?

alexanderivrii replied:
The implementation that you have reviewed followed the internal rustiq repo (the python wrapper around the rustiq-core code), which has two different functions for these tasks. 

But I completely agree that we can easily combine these two functions: in the unordered case we just create a "dummy" DAG without any edges. I believe the extra cost of checking the number of predecessors of a node for an edgeless DAG is completely negligible. 

Done (the commit will soon follow).

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+        if pauli_support_size == 0 {
+            // in case of an all-identity rotation, update global phase by subtracting
+            // the angle
+            global_phase = radd_param(global_phase, multiply_param(&angles[i], -1.0, py), py);
```
It might be worth it to only do the multiplication by -1 after the for loop since it's not for free (but here it likely doesn't matter much since you're only dealing with `Param::Float` and no `Param::ParameterExpression` which are very expensive to multiply)

alexanderivrii replied:
Hmm, the `synth_pauli_network_rustiq` function (and the underlying Rust code) work with parameterized angles. I have added a test for that. 

However, it does seem that our plugin interface for `PauliEvolutionGate`s does not accept parameterized angles. Why is this the case?

In the extremely unlikely case that we have a pauli network with multiple parametric all-identity gates, say with angles `a`, `b`, `c`, the suggestion is to replace the computation of the global phase from `0 + (-1) * a + (-1) * b + (-1) *c` to `0 + (-1) * (a + b + c)`?

Why are `ParameterExpression` more expensive to multiply than to add?

Cryoris replied:
> In the extremely unlikely case that we have a pauli network with multiple parametric all-identity gates, say with angles a, b, c, the suggestion is to replace the computation of the global phase from 0 + (-1) * a + (-1) * b + (-1) *c to 0 + (-1) * (a + b + c)?

Yes 🙂 

> Why are ParameterExpression more expensive to multiply than to add?

Both are expensive, so we should do it as little as possible. The multiplication by -1 is something we can optimize 🙂 

> However, it does seem that our plugin interface for PauliEvolutionGates does not accept parameterized angles. Why is this the case?

Hmm it should, maybe the type hint is wrong

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+
+/// Return the vector of Qiskit's gate corresponding to the given vector
+/// of Rustiq's Clifford gate.
+fn qiskit_clifford_gates(gates: &Vec<CliffordGate>) -> Vec<QiskitGate> {
```
Just a suggestion, since this is a transformation (I sort of expected this to return all Qiskit Clifford gates) 
```suggestion
fn to_qiskit_clifford_gates(gates: &Vec<CliffordGate>) -> Vec<QiskitGate> {
```

alexanderivrii replied:
Done (the commit will soon follow)

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+#[pyfunction]
+#[pyo3(signature = (num_qubits, pauli_network, optimize_count=true, preserve_order=true, upto_clifford=false, upto_phase=false, resynth_clifford_method=1))]
+pub fn pauli_network_synthesis(
+    py: Python,
```
If you want you could skip this argument and just get the GIL handle from the bound, 
```rust
let py = pauli_network.py();
```
instead 🙂

alexanderivrii replied:
nice! (but unless it's faster, let me leave this as is)

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+///     pauli network. In addition, the argument `upto_phase` would be ignored.
+/// * upto_phase: if `true`, the global phase of the returned circuit may differ
+///     from the global phase of the given pauli network. The argument is ignored
+///     when `upto_clifford` is `true`.
```
```suggestion
///     from the global phase of the given pauli network. The argument is set
///     to `true` when `upto_clifford` is `true`.
```

alexanderivrii replied:
Changed the docstring based on this suggestion. (The commit will soon follow).

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+/// * upto_phase: if `true`, the global phase of the returned circuit may differ
+///     from the global phase of the given pauli network. The argument is ignored
+///     when `upto_clifford` is `true`.
+/// * resynth_clifford_method: describes the strategy to synthesize the final
```
It would be nice for this to be an enum, whose names could reflect the method. See the following link for a pyo3 enum we can expose to Python space: https://stackoverflow.com/questions/76322128/pyo3-how-to-return-enums-to-python-module for an example.

alexanderivrii replied:
While I completely agree that using enums is a great idea, here specifically I am a bit reluctant to do this, as this is an option I don't expect 99% of users to ever experiment with (and that's 99% of the users who will want to experiment with Rustiq in the first place). I have considered leaving this option out, but I myself needed it to gauge the expected performance of switching to extended Clifford simulator. 

On a related note, I do think that using enum for specifying the optimization criterion (depth, count, fidelity, or something else) is also a good idea, but I don't think it makes sense to do this just for pauli network synthesis using Rustiq. Personally, I would like to see this as an additional argument for `transpile`.

Cryoris replied:
Let's leave it for a follow up then 👍🏻

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+            .downcast::<PyList>()?
+            .iter()
+            .map(|v| v.extract())
+            .collect::<PyResult<_>>()?;
```
It should be possible to directly extract a Vec, something like
```suggestion
        let qubits: Vec<u32> = tuple.get_item(1)?.extract()?;
```

alexanderivrii replied:
Wow! Just wow! After all the effort I got that to work, here you effortlessly come up with a beautiful one-liner. Fixed. The commit will appear soon.

Cryoris replied:
I had the exact same process with @raynelfss, so all the praise goes to him 😛

Cryoris commented on crates/accelerate/src/synthesis/evolution/mod.rs:
```diff
@@ -0,0 +1,439 @@
+    upto_phase: bool,
+    resynth_clifford_method: usize,
+) -> PyResult<CircuitData> {
+    let mut paulis: Vec<String> = Vec::new();
```
Could we use `Vec::with_capacity(pauli_network.len())` for both vectors here?

alexanderivrii replied:
Done.

Cryoris commented on qiskit/synthesis/evolution/pauli_network.py:
```diff
@@ -0,0 +1,80 @@
+    References:
+        1. Timothée Goubault de Brugière and Simon Martiel,
+           *Faster and shorter synthesis of Hamiltonian simulation circuits*,
+           `arXiv:2404.03280 [quant-ph] <https://arxiv.org/abs/2404.03280>`_
```
I think we typically don't include the category 🙂 
```suggestion
           `arXiv:2404.03280 <https://arxiv.org/abs/2404.03280>`_
```

alexanderivrii replied:
I think we do, or at least we do both (`grep` found a huge number of citations that explicitly mention `[quant-ph]`).

Cryoris commented on qiskit/synthesis/evolution/pauli_network.py:
```diff
@@ -0,0 +1,80 @@
+        upto_phase=upto_phase,
+        resynth_clifford_method=resynth_clifford_method,
+    )
+    circuit = QuantumCircuit._from_circuit_data(out)
```
It might be better to add the registers for consistency with the other Pauli evolution plugin 🙂 
```suggestion
    circuit = QuantumCircuit._from_circuit_data(out, add_regs=True)
```

alexanderivrii replied:
done (the commit will appear soon)

Cryoris commented on qiskit/transpiler/passes/synthesis/hls_plugins.py:
```diff
@@ -262,15 +262,46 @@
+      - :class:`~.PauliEvolutionSynthesisRustiq`
+      - all-to-all
+    * - ``"default"``
+      - :class:`~.PauliEvolutionSynthesisDefault`
```
Just a note: you could skip these as they just truncate the path (but there is no path)
```suggestion
      - :class:`.PauliEvolutionSynthesisRustiq`
      - all-to-all
    * - ``"default"``
      - :class:`.PauliEvolutionSynthesisDefault`
```

alexanderivrii replied:
I actually know this one, but for some reason all of our plugin tables have the "~", including the one you have recently added for MCMT. :)

Cryoris replied:
Oh then I was too quick to copy+paste 😂 Fair enough, let's clean this up some time later

Cryoris commented on qiskit/transpiler/passes/synthesis/hls_plugins.py:
```diff
@@ -1047,8 +1079,88 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
 
 class PauliEvolutionSynthesisDefault(HighLevelSynthesisPlugin):
-    """The default implementation calling the attached synthesis algorithm."""
+    """Synthesize a PauliEvolutionGate using the default synthesis algorithm.
```
Would it be nice to add the links (also below and in the other synthesis plugin)?
```suggestion
    """Synthesize a :class:`.PauliEvolutionGate` using the default synthesis algorithm.
```

alexanderivrii replied:
I have changed this for the two PauliEvolutionGate plugins. It would not be a lot of work to do this for all the other plugins as well, but possibly best after the release.

Cryoris commented on qiskit/transpiler/passes/synthesis/hls_plugins.py:
```diff
@@ -1047,8 +1079,88 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
+
+        algo = high_level_object.synthesis
+
+        if not hasattr(algo, "expand"):
```
Maybe this is safer?
```suggestion
        if not isinstance(algo, ProductFormula):
```

alexanderivrii replied:
I like this, this is indeed nicer. Done.

Cryoris commented on test/python/transpiler/test_high_level_synthesis.py:
```diff
@@ -2607,5 +2607,57 @@ def test_annotated_mcx(self):
+        self.assertIn("rustiq", supported_plugin_names)
+
+    @data("default", "rustiq")
+    def test_correctness(self, plugin_name):
```
To check correctness we'd have to compare `qct` to the matrix-based evolution of the Trotter steps, see https://github.com/Qiskit/qiskit/blob/2efb59ce4cd3c2d8721390648cdca6d2fd41f792/test/python/circuit/library/test_evolution_gate.py#L42. But since that's already tested for the default decomposition, we could just check whether the `rustiq` plugin equals the `default`.

alexanderivrii replied:
I believe that this is precisely what gets done internally when calculating `Operator(qc)`.

Cryoris commented on test/python/transpiler/test_high_level_synthesis.py:
```diff
@@ -2607,5 +2607,57 @@ def test_annotated_mcx(self):
+            hls_pass = HighLevelSynthesis(hls_config=hls_config)
+            qct = hls_pass(qc)
+            cnt_ops = qct.count_ops()
+            self.assertEqual(cnt_ops["cx"], 9)
```
Could we check that the decomposition is correct for the `upto_phase` decomposition? `Operator.equiv` should be the right method 🙂

alexanderivrii replied:
Done.

Cryoris commented on test/python/transpiler/test_high_level_synthesis.py:
```diff
@@ -2607,5 +2607,57 @@ def test_annotated_mcx(self):
+    @data("default", "rustiq")
+    def test_correctness(self, plugin_name):
+        """Test that all plugins return a correct Operator"""
+        op = SparsePauliOp(["XXX", "YYY", "IZZ", "XZY"], coeffs=[1, 2, 3, 4])
```
Could we add one test for the global phase (so with the `II...`) pauli?

alexanderivrii replied:
done

Cryoris commented on test/python/transpiler/test_high_level_synthesis.py:
```diff
@@ -2607,5 +2607,57 @@ def test_annotated_mcx(self):
+    @data("default", "rustiq")
+    def test_correctness(self, plugin_name):
+        """Test that all plugins return a correct Operator"""
+        op = SparsePauliOp(["XXX", "YYY", "IZZ", "XZY"], coeffs=[1, 2, 3, 4])
```
Could we also add a check for `preserve_order=False`?

alexanderivrii replied:
I don't know how to add a `correctness` test: since with `preserve_order=False` the operator might be completely different, but I have added a test that checks that the argument is applied.

Cryoris reviewed (commented):
Looks pretty good to me! I've left a bunch of comments but they are all very small 🙂

alexanderivrii committed:
moving the pauli network synthesis logic into a separate file

alexanderivrii committed:
some code review suggestions

alexanderivrii committed:
simplifying the code by merging the oredered and unorderd version of rotation injection

alexanderivrii committed:
more review comments

alexanderivrii committed:
adding python tests

alexanderivrii committed:
more code review suggestions

alexanderivrii committed:
more review comments

alexanderivrii committed:
more review comments

alexanderivrii committed:
test for preserve_order

alexanderivrii commented:
Thanks @Cryoris for a thorough review! I think I have addressed all of your review comments (and in a few cases explained why I don't want to address it). To keep track of your comments, I was marking them as resolved as I was going through them, so you will probably want to unresolve them to see if a deeper discussion is required.

alexanderivrii committed:
lint

alexanderivrii committed:
upgrading rustiq-core to 0.0.10

alexanderivrii committed:
clippy: removing mutable ref

alexanderivrii committed:
Improving PauliEvolution synthesis tests.

Making sure that the number of rotation gates in the synthesized
circuit equals the number of non-trivial Pauli rotation gates.

alexanderivrii committed:
Merge branch 'main' into integrate_rustiq_plugin

alexanderivrii committed:
documentation fixes after the merge

Cryoris reviewed (approved):
LGTM thanks for the updates!
